File: src/jvm/de/rub/bi/inf/openbimrl/functions/ifc/GetQuantity.java
package de.rub.bi.inf.openbimrl.functions.ifc;

import de.rub.bi.inf.openbimrl.NodeProxy;
import de.rub.bi.inf.openbimrl.engine.ifc.IIFCModel;
import de.rub.bi.inf.openbimrl.functions.AbstractFunction;
import de.rub.bi.inf.openbimrl.helper.IfcUtilities;

import java.util.ArrayList;
import java.util.Collection;

/**
 * Retrieves the quantity value of a specific entity.
 *
 * @author Marcel Stepien
 */
public class GetQuantity extends AbstractFunction {

    public GetQuantity(NodeProxy nodeProxy) {
        super(nodeProxy);
    }

    @Override
    public void execute(IIFCModel ifcModel) {

/*        Collection<Object> objects = new ArrayList();
        Object input0 = getInput(0);
        if (input0 instanceof Collection<?>) {
            objects = (Collection<Object>) input0;
        } else {
            objects.add(input0);
        }


        String setName = getInput(1);
        String propertyName = getInput(2);

        ArrayList<Object> resultValues = new ArrayList<>();

        for (Object o : objects) {
            IfcObjectDefinition.Ifc4 ifcObjectDefinition = (IfcObjectDefinition.Ifc4) o;

            IfcPhysicalQuantity physicalQuantity = IfcUtilities.Ifc4.getElementQuantity(ifcObjectDefinition, setName, propertyName);

            if (physicalQuantity == null) {
                resultValues.add(null);
            } else if (physicalQuantity instanceof IfcQuantityArea.Ifc4) {
                IfcQuantityArea.Ifc4 quantityArea =
                        (IfcQuantityArea.Ifc4) physicalQuantity;
                resultValues.add(quantityArea.getAreaValue().getValue());
            } else if (physicalQuantity instanceof IfcQuantityLength.Ifc4) {
                resultValues.add(((IfcQuantityLength.Ifc4) physicalQuantity).getLengthValue().getValue());
            } else {

                //FIXME other possibilities
                System.err.println(physicalQuantity.getClassName() + "not covered yet");
                resultValues.add(null);

            }

        }

        if (resultValues.size() == 1) {
            setResult(0, resultValues.get(0));
        } else {
            setResult(0, resultValues);
        }*/
    }

}

File: src/main/jvm/de/rub/bi/inf/openbimrl/functions/ifc/GetQuantity.kt
package de.rub.bi.inf.openbimrl.functions.ifc

import de.rub.bi.inf.nativelib.IfcPointer
import de.rub.bi.inf.openbimrl.NodeProxy
import de.rub.bi.inf.openbimrl.functions.AbstractFunction
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionInput
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionOutput
import de.rub.bi.inf.openbimrl.functions.annotations.OpenBIMRLFunction

@OpenBIMRLFunction(description = "Retrieves the quantity value of a specific entity.")
class GetQuantity(nodeProxy: NodeProxy) : AbstractFunction(nodeProxy) {

    @FunctionInput(0, name = "IfcElement", collectionType = IfcPointer::class)
    lateinit var ifcElements: Collection<IfcPointer>

    @FunctionInput(1)
    lateinit var quantitySetName: String

    @FunctionInput(2)
    lateinit var quantityName: String

    @FunctionOutput(0)
    var value: List<Double>? = null

    override fun execute() {
        value = ifcElements.filter { element ->
            element.quantities.containsKey(quantitySetName) &&
                element.quantities[quantitySetName]!!.containsKey(quantityName)
        }.map { element ->
            element.quantities[quantitySetName]!![quantityName]!!
        }
    }
}
